crates/sui-types/src/messages.rs

```diff
@@ -231,65 +231,298 @@ pub enum SingleTransactionKind {
+        // because they must all be on-chain in order for the package to publish.
+        // All authorities must have the same view of those dependencies in order
+        // to achieve consistent publish results.
+        let compiled_modules = modules
+            .iter()
+            .filter_map(|bytes| match CompiledModule::deserialize(bytes) {
+                Ok(m) => Some(m),
+                // We will ignore this error here and simply let latter execution
+                // to discover this error again and fail the transaction.
+                // It's preferable to let transaction fail and charge gas when
+                // malformed package is provided.
+                Err(_) => None,
+            })
+            .collect::<Vec<_>>();
+        Transaction::input_objects_in_compiled_modules(&compiled_modules)
+    }
+
+    fn input_objects(&self) -> Vec<InputObjectKind> {
+        match self {
+            Command::Publish(modules) => Self::publish_command_input_objects(modules),
+            Command::MoveCall(c) => c.input_objects(),
+            Command::TransferObjects(_, _)
+            | Command::SplitCoin(_, _)
+            | Command::MergeCoins(_, _) => vec![],
+        }
+    }
+
+    fn validity_check(&self) -> SuiResult {
+        match self {
+            Command::MoveCall(call) => {
+                let is_blocked = BLOCKED_MOVE_FUNCTIONS.contains(&(
+                    call.package,
+                    call.module.as_str(),
+                    call.function.as_str(),
+                ));
+                fp_ensure!(!is_blocked, SuiError::BlockedMoveFunction);
+            }
+            Command::TransferObjects(_, _)
+            | Command::SplitCoin(_, _)
+            | Command::MergeCoins(_, _)
+            | Command::Publish(_) => (),
+        };
+        Ok(())
+    }
+}
+
+fn write_sep<T: Display>(
+    f: &mut Formatter<'_>,
+    items: impl IntoIterator<Item = T>,
+    sep: &str,
+) -> std::fmt::Result {
+    let mut xs = items.into_iter().peekable();
+    while let Some(x) = xs.next() {
+        if xs.peek().is_some() {
+            write!(f, "{sep}")?;
+        }
+        write!(f, "{x}")?;
+    }
+    Ok(())
+}
+
+impl ProgrammableTransaction {
+    pub fn input_objects(&self) -> SuiResult<Vec<InputObjectKind>> {
+        let ProgrammableTransaction { inputs, commands } = self;
+        let input_arg_objects = inputs
+            .iter()
+            .flat_map(|arg| arg.input_objects())
+            .collect::<Vec<_>>();
+        let mut used = HashSet::new();
+        if !input_arg_objects.iter().all(|o| used.insert(o.object_id())) {
+            return Err(SuiError::DuplicateObjectRefInput);
+        }
+        Ok(input_arg_objects
+            .into_iter()
+            .chain(commands.iter().flat_map(|command| command.input_objects()))
+            .collect())
+    }
+
+    fn validity_check(&self) -> SuiResult {
+        if !cfg!(test) {
+            return Err(SuiError::Unknown(
```

Comment:
Does the error kind matter here for compatibility/deployment? Should I give `SuiError::InvalidDecoding` or something instead?